src/tests/tribol_enzyme_merged_jacobian.cpp

```diff
@@ -0,0 +1,656 @@
+        }
+        mesh1_elem_map.push_back( e );
+      }
+    }
+
+    // register Tribol mesh data
+    mfem::GridFunction coords( &fe_space );
+    if ( parent_coords == nullptr ) {
+      coords = ref_coords;
+    } else {
+      submesh.Transfer( *parent_coords, coords );
+    }
+    auto coords_ptr = coords.Read();
+    constexpr auto mesh0_id = 0;
+    registerMesh( mesh0_id, num_contact_els_mesh0, fe_space.GetNDofs(), mesh0_conn.data(), LINEAR_QUAD, coords_ptr,
+                  coords_ptr + fe_space.GetNDofs(), coords_ptr + 2 * fe_space.GetNDofs() );
+    constexpr auto mesh1_id = 1;
+    registerMesh( mesh1_id, num_contact_els_mesh1, fe_space.GetNDofs(), mesh1_conn.data(), LINEAR_QUAD, coords_ptr,
+                  coords_ptr + fe_space.GetNDofs(), coords_ptr + 2 * fe_space.GetNDofs() );
+    // mesh will move when we do finite differencing
+    auto ref_coords_ptr = ref_coords.Read();
+    registerNodalReferenceCoords( mesh0_id, ref_coords_ptr, ref_coords_ptr + fe_space.GetNDofs(),
+                                  ref_coords_ptr + 2 * fe_space.GetNDofs() );
+    registerNodalReferenceCoords( mesh1_id, ref_coords_ptr, ref_coords_ptr + fe_space.GetNDofs(),
+                                  ref_coords_ptr + 2 * fe_space.GetNDofs() );
+
+    // create MFEM force linear form
+    mfem::LinearForm force( &fe_space );
+    force = 0.0;
+    auto force_ptr = force.Write();
+    registerNodalResponse( mesh0_id, force_ptr, force_ptr + fe_space.GetNDofs(), force_ptr + 2 * fe_space.GetNDofs() );
+    registerNodalResponse( mesh1_id, force_ptr, force_ptr + fe_space.GetNDofs(), force_ptr + 2 * fe_space.GetNDofs() );
+
+    // create MFEM pressure grid function
+    mfem::FiniteElementSpace fe_space_scalar( &submesh, &fe_coll, 1 );
+    mfem::GridFunction fake_pressure( &fe_space_scalar );
+    fake_pressure = 1.0;
+    if ( pressure == nullptr ) {
+      pressure = &fake_pressure;
+    }
+    auto pressure_ptr = pressure->Read();
+    registerMortarPressures( mesh1_id, pressure_ptr );
+
+    // create MFEM gap linear form
+    mfem::LinearForm gap( &fe_space_scalar );
+    gap = 0.0;
+    auto gap_ptr = gap.Write();
+    registerMortarGaps( mesh1_id, gap_ptr );
+
+    // register Tribol coupling scheme
+    constexpr auto cs_id = 0;
+    registerCouplingScheme( cs_id, mesh0_id, mesh1_id, ContactMode::SURFACE_TO_SURFACE, ContactCase::NO_CASE,
+                            ContactMethod::SINGLE_MORTAR, ContactModel::FRICTIONLESS,
+                            EnforcementMethod::LAGRANGE_MULTIPLIER, BinningMethod::BINNING_GRID,
+                            ExecutionMode::Sequential );
+    setLagrangeMultiplierOptions( cs_id, ImplicitEvalMode::MORTAR_RESIDUAL_JACOBIAN, SparseMode::MFEM_ELEMENT_DENSE );
+    enableEnzyme( cs_id, true );
+
+    // calc element force, gap, and Jacobians
+    constexpr auto cycle = 1;
+    constexpr auto t = 1.0;
+    auto dt = 1.0;
+    update( cycle, t, dt );
+
+    // get element Jacobian contributions and assemble
```

Comment:
Are these enzyme jacobian contributions, or are these the hand coded approximate jacobian contributions? That is, if enzyme is enabled, does "getBlockJ()" return enzyme Jacobians and if enzyme is NOT enabled, then "getBlockJ()" will default to the hand coded approximate jacobian contributions?

Reply:
These are enzyme Jacobian contributions.  Yes, `getBlockJ()` will return approximate Jacobian calculations when enzyme is off.